### ISSUE TITLE
Kubeadm-related privileges required even for non-kubeadm clusters

### ISSUE THREAD

**Issue by seeker89** [projectcalico/node#489]
Hi,

I've run into an issue during upgrading calico to v3.13.3 from v3.10.0, where (on a cluster not bootstrapped with `kubeadm`) I didn't give `calico-node` the `kubeadm` privs to read `configmaps` [like you recommend in the manifests](https://docs.projectcalico.org/manifests/calico.yaml):

```yaml
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: calico-node
rules:
(...)
  # Pod CIDR auto-detection on kubeadm needs access to config maps.
  - apiGroups: [""]
    resources:
      - configmaps
    verbs:
      - get
```

My thinking was - since I don't use `kubeadm`, I should skip them.

## Expected Behavior

Log a message and start up correctly, assuming it's not a `kubeadm` cluster.

## Current Behavior

The code below crashes with a generic error code 1, and this log message:

```
2020-05-06 11:59:16.853 [ERROR][10] startup.go 146: failed to query kubeadm's config map error=configmaps "kubeadm-config" is forbidden: User "system:serviceaccount:kube-system:calico-node" cannot get resource "configmaps" in API group "" in the namespace "kube-system"
2020-05-06 11:59:16.853 [WARNING][10] startup.go 1214: Terminating
```

https://github.com/projectcalico/node/blob/93d7706db41f74a75438595ec46f500f7f822129/pkg/startup/startup.go#L140-L150

## Possible Solution

If error is 403 Forbidden - then assume it's not a `kubeadm` cluster.

## Steps to Reproduce (for bugs)

1. Remove the priv described at the top
2. Restart the calico-node containers
3. Watch them crash and burn

## Your Environment

* calico/node version: v3.13.3
* Operating System and version: Ubuntu bionic

**Comment by caseydavenport** [comment]
@seeker89 I think the enhancement you laid out seems like a sensible one. We could check the error type to see if it was an unauthorized connection, and if so assume it's because we're not on kubeadm.

**Comment by caseydavenport** [comment]
> I didn't give calico-node the kubeadm privs to read configmaps like you recommend in the manifests:

I think it's worth pointing out, however, that the manifests there are not a recommendation per-se. The dev team treats those manifests as code in the product would be, and so the assumption is that if you are using the Calico v3.X.Y binaries you are using the matching Calico v3.X.Y manifests.

**Comment by seeker89** [comment]
Thanks for your response! In this case, I think the best course of action would be to:

1) add this extra test when anyone has a spare cycle
2) make it hard to miss in the documentation, that none of the stuff from the manifests is optional

@caseydavenport do you agree?

**Comment by DiCanio** [comment]
@seeker89 

I'm with you on the first point since a cluster can operate without kubeadm being around. So there's no point in adding unnecessary resource access rights.

However, pointing out in the documentation that manifests must be used as is could be quite misleading regarding optional parts of it. If something is optional AND handled as such by the code then manifests should reflect that fact. I would reckon that commenting out optional parts in the manifest is the way to go. Just leave it to the user to enable it since he knows how his cluster is set up. If this is not possible due to whatever reasons then a kubeadm check at least resolves the initial issue.

**Comment by seeker89** [comment]
> However, pointing out in the documentation that manifests must be used as is could be quite misleading regarding optional parts of it. If something is optional AND handled as such by the code then manifests should reflect that fact. I would reckon that commenting out optional parts in the manifest is the way to go. Just leave it to the user to enable it since he knows how his cluster is set up. If this is not possible due to whatever reasons then a kubeadm check at least resolves the initial issue.

True, so let me rephrase that proposal:

1. add this extra test when anyone has a spare cycle
2. make it easy to understand which parts of the manifest are required, and which ones are required

**Comment by caseydavenport** [comment]
On point (2), this is something I've thought a little bit about - the Kubernetes manifests make it really hard from a maintainer's perspective to represent what is a "supported feature" and what is "required to make the code work correctly". It's all very... hard to pin down, and as a result leads to a proliferation of both-hard-to-write-and-understand documentation and dissimilar deployments across different users, resulting in more difficult troubleshooting and very different experiences for different users. The upside, of course, is limitless flexibility for those who want it. 

It's one of the reasons I'm currently excited about what we're exploring with https://github.com/tigera/operator as a means of managing the lifecycle of a Calico install - it lets us write an API around the product to help ensure a certain quality and consistency of service, if you will, for Calico users.

**Comment by ovesh** [comment]
Looks like this was already done in https://github.com/projectcalico/node/pull/541
